File: modules/virtual-people-core-serving/src/test/kotlin/org/wfanet/virtualpeople/core/model/ConditionalMergeImplTest.kt
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.core.model

import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.wfanet.virtualpeople.common.BranchNodeKt.attributesUpdater
import org.wfanet.virtualpeople.common.ConditionalMergeKt.conditionalMergeNode
import org.wfanet.virtualpeople.common.FieldFilterProto.Op
import org.wfanet.virtualpeople.common.conditionalMerge
import org.wfanet.virtualpeople.common.fieldFilterProto
import org.wfanet.virtualpeople.common.labelerEvent

@RunWith(JUnit4::class)
class ConditionalMergeImplTest {

  @Test
  fun `no node should throw`() {
    val config = attributesUpdater {
      conditionalMerge = conditionalMerge { passThroughNonMatches = false }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("No nodes in ConditionalMerge"))
  }

  @Test
  fun `no condition should throw`() {
    val config = attributesUpdater {
      conditionalMerge = conditionalMerge {
        nodes.add(
          conditionalMergeNode { update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_1" } }
        )
        passThroughNonMatches = false
      }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("No condition in the node"))
  }

  @Test
  fun `invalid condition should throw`() {
    val config = attributesUpdater {
      conditionalMerge = conditionalMerge {
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto { op = Op.EQUAL }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_1" }
          }
        )
        passThroughNonMatches = false
      }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("Name must be set."))
  }

  @Test
  fun `no update should throw`() {
    val config = attributesUpdater {
      conditionalMerge = conditionalMerge {
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_1"
            }
          }
        )
        passThroughNonMatches = false
      }
    }
    val error = assertFailsWith<IllegalStateException> { AttributesUpdaterInterface.build(config) }
    assertTrue(error.message!!.contains("No update in the node"))
  }

  @Test
  fun `update event successfully`() {
    val config = attributesUpdater {
      conditionalMerge = conditionalMerge {
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_1"
            }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_1" }
          }
        )
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_2"
            }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_2" }
          }
        )
        passThroughNonMatches = false
      }
    }
    val updater = AttributesUpdaterInterface.build(config)

    /** Matches the 1st node, update the event. */
    val eventBuilder1 = labelerEvent { personCountryCode = "COUNTRY_1" }.toBuilder()
    updater.update(eventBuilder1)
    assertEquals("UPDATED_COUNTRY_1", eventBuilder1.personCountryCode)

    /** Matches the 2st node, update the event. */
    val eventBuilder2 = labelerEvent { personCountryCode = "COUNTRY_2" }.toBuilder()
    updater.update(eventBuilder2)
    assertEquals("UPDATED_COUNTRY_2", eventBuilder2.personCountryCode)
  }

  @Test
  fun `no matching not pass should throw`() {
    val config = attributesUpdater {
      conditionalMerge = conditionalMerge {
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_1"
            }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_1" }
          }
        )
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_2"
            }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_2" }
          }
        )
        passThroughNonMatches = false
      }
    }
    val updater = AttributesUpdaterInterface.build(config)

    /** No matching. Returns error status as pass_through_non_matches is false. */
    val eventBuilder = labelerEvent { personCountryCode = "COUNTRY_3" }.toBuilder()
    val exception = assertFailsWith<IllegalStateException> { updater.update(eventBuilder) }
    assertTrue(exception.message!!.contains("No node matching"))
    assertEquals("COUNTRY_3", eventBuilder.personCountryCode)
  }

  @Test
  fun `no matching not pass by default should throw`() {
    val config = attributesUpdater {
      conditionalMerge = conditionalMerge {
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_1"
            }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_1" }
          }
        )
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_2"
            }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_2" }
          }
        )
      }
    }
    val updater = AttributesUpdaterInterface.build(config)

    /** No matching. Returns error status as pass_through_non_matches is false by default. */
    val eventBuilder = labelerEvent { personCountryCode = "COUNTRY_3" }.toBuilder()
    val exception = assertFailsWith<IllegalStateException> { updater.update(eventBuilder) }
    assertTrue(exception.message!!.contains("No node matching"))
    assertEquals("COUNTRY_3", eventBuilder.personCountryCode)
  }

  @Test
  fun `no matching pass should do nothing`() {
    val config = attributesUpdater {
      conditionalMerge = conditionalMerge {
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_1"
            }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_1" }
          }
        )
        nodes.add(
          conditionalMergeNode {
            condition = fieldFilterProto {
              op = Op.EQUAL
              name = "person_country_code"
              value = "COUNTRY_2"
            }
            update = labelerEvent { personCountryCode = "UPDATED_COUNTRY_2" }
          }
        )
        passThroughNonMatches = true
      }
    }
    val updater = AttributesUpdaterInterface.build(config)

    /** No matching. Do nothing as pass_through_non_matches is true. */
    val eventBuilder = labelerEvent { personCountryCode = "COUNTRY_3" }.toBuilder()
    updater.update(eventBuilder)
    assertEquals("COUNTRY_3", eventBuilder.personCountryCode)
  }
}

File: modules/virtual-people-core-serving/src/test/cc/wfa/virtual_people/core/model/conditional_merge_impl_test.cc
// Copyright 2021 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "common_cpp/testing/status_macros.h"
#include "common_cpp/testing/status_matchers.h"
#include "gmock/gmock.h"
#include "google/protobuf/text_format.h"
#include "gtest/gtest.h"
#include "wfa/virtual_people/common/model.pb.h"
#include "wfa/virtual_people/core/model/attributes_updater.h"

namespace wfa_virtual_people {
namespace {

using ::wfa::IsOk;
using ::wfa::StatusIs;

TEST(ConditionalMergeImplTest, TestNoNodes) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_merge { pass_through_non_matches: false }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalMergeImplTest, TestNoCondition) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_merge {
          nodes { update { person_country_code: "UPDATED_COUNTRY_1" } }
          pass_through_non_matches: false
        }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalMergeImplTest, TestInvalidCondition) {
  // Name and value must be set for EQUAL field filter. So the build of the
  // field filter condition of the 1st node will fail, thus the buid of the
  // conditional merge will fail.
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_merge {
          nodes {
            condition { op: EQUAL }
            update { person_country_code: "UPDATED_COUNTRY_1" }
          }
          pass_through_non_matches: false
        }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalMergeImplTest, TestNoUpdate) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_merge {
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_1"
            }
          }
          pass_through_non_matches: false
        }
      )pb",
      &config));
  EXPECT_THAT(AttributesUpdaterInterface::Build(config).status(),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
}

TEST(ConditionalMergeImplTest, TestUpdateEvents) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_merge {
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_1"
            }
            update { person_country_code: "UPDATED_COUNTRY_1" }
          }
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_2"
            }
            update { person_country_code: "UPDATED_COUNTRY_2" }
          }
          pass_through_non_matches: false
        }
      )pb",
      &config));
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<AttributesUpdaterInterface> updater,
                       AttributesUpdaterInterface::Build(config));

  // Matches the 1st node, update the event.
  LabelerEvent event_1;
  event_1.set_person_country_code("COUNTRY_1");
  EXPECT_THAT(updater->Update(event_1), IsOk());
  EXPECT_EQ(event_1.person_country_code(), "UPDATED_COUNTRY_1");

  // Matches the 2nd node, update the event.
  LabelerEvent event_2;
  event_2.set_person_country_code("COUNTRY_2");
  EXPECT_THAT(updater->Update(event_2), IsOk());
  EXPECT_EQ(event_2.person_country_code(), "UPDATED_COUNTRY_2");
}

TEST(ConditionalMergeImplTest, TestNoMatchingNotPass) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_merge {
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_1"
            }
            update { person_country_code: "UPDATED_COUNTRY_1" }
          }
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_2"
            }
            update { person_country_code: "UPDATED_COUNTRY_2" }
          }
          pass_through_non_matches: false
        }
      )pb",
      &config));
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<AttributesUpdaterInterface> updater,
                       AttributesUpdaterInterface::Build(config));

  // No matching. Returns error status as pass_through_non_matches is false.
  LabelerEvent event;
  event.set_person_country_code("COUNTRY_3");
  EXPECT_THAT(updater->Update(event),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
  EXPECT_EQ(event.person_country_code(), "COUNTRY_3");
}

TEST(ConditionalMergeImplTest, TestNoMatchingNotPassByDefault) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_merge {
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_1"
            }
            update { person_country_code: "UPDATED_COUNTRY_1" }
          }
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_2"
            }
            update { person_country_code: "UPDATED_COUNTRY_2" }
          }
        }
      )pb",
      &config));
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<AttributesUpdaterInterface> updater,
                       AttributesUpdaterInterface::Build(config));

  // No matching. pass_through_non_matches is false by default.
  LabelerEvent event;
  event.set_person_country_code("COUNTRY_3");
  EXPECT_THAT(updater->Update(event),
              StatusIs(absl::StatusCode::kInvalidArgument, ""));
  EXPECT_EQ(event.person_country_code(), "COUNTRY_3");
}

TEST(ConditionalMergeImplTest, TestNoMatchingPass) {
  BranchNode::AttributesUpdater config;
  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(
      R"pb(
        conditional_merge {
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_1"
            }
            update { person_country_code: "UPDATED_COUNTRY_1" }
          }
          nodes {
            condition {
              op: EQUAL
              name: "person_country_code"
              value: "COUNTRY_2"
            }
            update { person_country_code: "UPDATED_COUNTRY_2" }
          }
          pass_through_non_matches: true
        }
      )pb",
      &config));
  ASSERT_OK_AND_ASSIGN(std::unique_ptr<AttributesUpdaterInterface> updater,
                       AttributesUpdaterInterface::Build(config));

  // No matching. Returns OK status as pass_through_non_matches is true.
  LabelerEvent event;
  event.set_person_country_code("COUNTRY_3");
  EXPECT_THAT(updater->Update(event), IsOk());
  EXPECT_EQ(event.person_country_code(), "COUNTRY_3");
}

}  // namespace
}  // namespace wfa_virtual_people
